src/Data/Functor/Foldable/TH.hs

```diff
@@ -1,4 +1,13 @@
 {-# LANGUAGE CPP, PatternGuards, Rank2Types #-}
+-- This OPTIONS_GHC line is a workaround for
+-- https://gitlab.haskell.org/ghc/ghc/-/issues/18320, a bug which only occurs
+-- when running specific TemplateHaskell code while both profiling and
+-- optimisations are enabled. The code in this file triggers the bug, so until
+-- it is fixed, we work around the issue by disabling optimisations in this
+-- file. The code in this file only runs at compile-time, the code _generated_
+-- by makeBaseFunctor will still get optimized if the file which calls
+-- makeBaseFunctor is optimized.
```

Comment:
Here's how I checked that claim. I created a larger variant of our first test, and split it into three files:

```haskell
module Datatype where

data Expr a
    = Lit a
    | Add (Expr a) (Expr a)
    | Expr a :* [Expr a]
  deriving (Show)
```

```haskell
module TH where

import Data.Functor.Foldable.TH
import Datatype

makeBaseFunctor ''Expr
```

```haskell
module Main where

import Data.Functor.Foldable
import Datatype
import TH

mkDeepTree :: Natural -> Expr Int
mkDeepTree = cata mkDeepTreeF

mkDeepTreeF :: Maybe (Expr Int) -> Expr Int
mkDeepTreeF Nothing
  = Lit 1
mkDeepTreeF (Just e)
  = Add e (e :* [e])

expr1 :: Expr Int
expr1 = mkDeepTree 18

main :: IO ()
main = do
    14 @=? cata evalAlg expr1
  where
    evalAlg :: ExprF Int Int -> Int
    evalAlg (LitF x)   = x
    evalAlg (AddF x y) = x + y
    evalAlg (x :*$ y) = foldl' (*) x y
```

I then experimented with various combinations, adding `{-# GHC_OPTIONS -O0 #-}` or `{-# GHC_OPTIONS -O2 #-}` to the `Data.Functor.Foldable.TH` (def) and `TH` (call) modules, and running the tests with either `cabal test --enable-profiling` or just `cabal test`. I made sure to run the command several times before capturing the duration, in order to make sure that I was not measuring the compilation time and that all caches were warm. I got the following results:

|                   | no profiling | with profiling |
|-------------------|--------------|----------------|
| *def O0, call O0* |       37.819 |         79.675 |
| *def O2, call O0* |       37.345 |         78.796 |
| *def O0, call O2* |       16.899 |         37.280 |
| *def O2, call O2* |       17.081 |        ghc bug |

I think this confirms that

1. Setting `-O0` in `Data.Functor.Foldable.TH` (def) doesn't affect the runtime performance, only the compile time performance (which is not shown in this chart).
2. Setting `-O2` at the `makeBaseFunctor` call site does cause the generated code to be optimized.